nep141-token-convertor-contract/src/conversion_pool.rs

```diff
@@ -0,0 +1,313 @@
+        let mut pool = self.internal_get_pool(&pool_id).expect("No such pool");
+        let r = f(&mut pool);
+        self.internal_save_pool(pool_id, &pool.into());
+        r
+    }
+
+    #[private]
+    pub(crate) fn internal_get_pool(&self, pool_id: &PoolId) -> Option<ConversionPool> {
+        // return self.pools.get(&pool_id)
+        return self.pools.get(pool_id).map(|pool| pool.into_current());
+    }
+
+    #[private]
+    pub(crate) fn internal_save_pool(&mut self, pool_id: PoolId, pool: &VPool) {
+        self.pools.insert(&pool_id, &pool);
+        // self.pools.replace(pool_id, &pool);
+    }
+}
+
+#[near_bindgen]
+impl PoolCreatorAction for TokenConvertor {
+    #[payable]
+    fn create_pool(
+        &mut self,
+        in_token: AccountId,
+        out_token: AccountId,
+        is_reversible: bool,
+        in_token_rate: u32,
+        out_token_rate: u32,
+    ) -> u32 {
+        self.assert_contract_is_not_paused();
+        assert!(
+            !in_token.eq(&out_token),
+            "You can't create pool for same token"
+        );
+        self.assert_create_pool_deposit_amount();
+        self.assert_token_in_whitelist(&in_token);
+        self.assert_token_in_whitelist(&out_token);
+        assert_eq!(
+            self.whitelisted_tokens.get(&in_token).unwrap().decimals,
+            self.whitelisted_tokens.get(&out_token).unwrap().decimals,
+            "tokens in a pool should have same decimals."
+        );
+        let id = self.internal_assign_pool_id();
+        self.pools.insert(
+            &id,
+            &VPool::Current(ConversionPool::new(
+                id.clone(),
+                env::predecessor_account_id(),
+                in_token.clone(),
+                out_token.clone(),
+                is_reversible,
+                in_token_rate,
+                out_token_rate,
+                U128(env::attached_deposit()),
+            )),
+        );
+        id
+    }
+
+    #[payable]
+    fn withdraw_token_in_pool(&mut self, pool_id: PoolId, token_id: AccountId, amount: U128) {
+        self.assert_contract_is_not_paused();
+        assert_one_yocto();
+        self.internal_use_pool(pool_id, |pool| {
+            assert_eq!(
+                pool.creator,
+                env::predecessor_account_id(),
+                "Only creator can remove liquidity."
+            );
+            assert!(
+                token_id == pool.in_token || token_id == pool.out_token,
+                "Illegal token id {}",
+                token_id
+            );
+            if token_id == pool.in_token {
+                // if fail, it should panic
+                pool.withdraw_in_token(amount.0);
+            } else {
+                pool.withdraw_out_token(amount.0);
+            }
+        });
+        // pool should finish withdraw here
+        self.internal_send_tokens(&env::predecessor_account_id(), &token_id, amount.0);
+    }
+
+    #[payable]
+    fn delete_pool(&mut self, pool_id: PoolId) {
```

Comment:
There are 3 possible promises which can be created in this function. I'm not sure what will happen if the prepaid gas can only fulfill the first one or two promise(s). Need to double check.
I recommend to check the prepaid gas before actual process.

Reply:
I have changed logic in `delete_pool`.Now requiring withdraw all `in_token` and `out_token` in pool before deleting pool.And admin can also execute withdraw token for creator now.